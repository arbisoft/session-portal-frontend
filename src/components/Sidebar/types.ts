export type SidebarProps = {
  handleSidebarToggle?: VoidFunction;
};
