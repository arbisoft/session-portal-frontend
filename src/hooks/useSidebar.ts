import { useEffect, useState } from "react";

const useSidebar = () => {
  const [sidebarItems, setSidebarItems] = useState<string[]>([]);

  useEffect(() => {
    const onGetSidebarItems = async () => {
      //   TODO: Remove static code and get items from backend
      setSidebarItems(["All", "All-Hands Meetings", "AMA Sessions", "Product Showcase", "Nurture Sessions", "Security Session"]);
    };
    onGetSidebarItems();
  }, []);

  return { sidebarItems };
};
export default useSidebar;
