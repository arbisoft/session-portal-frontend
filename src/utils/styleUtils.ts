export const shouldForwardProp = (propName: string) => !propName.startsWith("$");

export const pxToRem = (px: number, base = 16): string => {
  return `${(px * 1.25) / base}rem`;
};
