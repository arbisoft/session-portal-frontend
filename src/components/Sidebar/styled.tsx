import Box from "@mui/material/Box";
import { chipClasses } from "@mui/material/Chip";
import MenuItem from "@mui/material/MenuItem";
import Stack from "@mui/material/Stack";
import { styled, css } from "@mui/material/styles";
import Typography from "@mui/material/Typography";

import { pxToRem, shouldForwardProp } from "@/utils/styleUtils";

export const SidebarContainer = styled(Box, {
  name: "SidebarContainer",
})(() => {
  return css`
    height: 100%;
    width: 100%;
  `;
});

export const MenuStack = styled(Stack, {
  name: "MenuStack",
})(() => {
  return css`
    flex-direction: column;
    height: 100%;
    gap: 8px;
  `;
});

export const StyledMenuItem = styled(MenuItem, {
  name: "MenuItem",
  shouldForwardProp,
})(() => {
  return css`
    border-radius: 6px;
    cursor: pointer;
    flex-direction: row;
    gap: 10px;
    padding: 12px;
  `;
});

export const Text = styled(Typography, {
  name: "Text",
})(() => {
  return css`
    font-size: ${pxToRem(14)};
    line-height: 12px;
    font-weight: 500;
    letter-spacing: 0.4px;
  `;
});

export const TagsContainer = styled(Box, {
  name: "TagsContainer",
})(
  () => css`
    display: flex;
    flex-wrap: wrap;
    gap: 5px;
    padding-top: 30px;

    .${chipClasses.root} {
      border-radius: 6px;
      cursor: pointer;
      text-transform: capitalize;

      .${chipClasses.icon} {
        margin: 0;
      }
    }
  `
);
