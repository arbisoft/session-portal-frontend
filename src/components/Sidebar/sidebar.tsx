import React from "react";

import { faker } from "@faker-js/faker";
import Box from "@mui/material/Box";
import Chip from "@mui/material/Chip";
import MenuItem from "@mui/material/MenuItem";
import MenuList from "@mui/material/MenuList";
import Skeleton from "@mui/material/Skeleton";
import Image from "next/image";
import { useSearchParams } from "next/navigation";

import useNavigation from "@/hooks/useNavigation";
import useSidebar from "@/hooks/useSidebar";
import { runsOnServerSide } from "@/services/runs-on-server-side/runs-on-server-side";

import { StyledMenuItem, MenuStack, SidebarContainer, TagsContainer, Text } from "./styled";

const loadingTags: string[] = Array(10)
  .fill("")
  .map(() => faker.lorem.words(1));

const Sidebar = () => {
  const { navigateTo } = useNavigation();
  const searchParams = useSearchParams();

  const { arePlaylistsLoading, tags, playlists } = useSidebar();

  const tag = searchParams?.get("tag");
  const playlist = searchParams?.get("playlist");

  const onResetFilters = () => {
    if (!runsOnServerSide) window.dispatchEvent(new Event("reset-filter"));
  };

  return (
    <SidebarContainer data-testid="sidebar-container">
      <MenuStack>
        {arePlaylistsLoading ? (
          <MenuList data-testid="loading">
            {loadingTags?.map((item) => (
              <MenuItem key={item}>
                <Box key={item} display="flex" justifyContent="space-between" alignItems={"center"} width="90%" mb={1}>
                  <Skeleton variant="rounded" width="15%" height={20} />
                  <Skeleton width="76%" height={25} />
                </Box>
              </MenuItem>
            ))}
          </MenuList>
        ) : (
          <MenuList>
            <StyledMenuItem selected={!playlist && !tag} onClick={() => navigateTo("videos")} data-testid={"sidebar-item-All"}>
              <Image src="/assets/images/sidebar-item-icon.svg" alt="All videos" width={18} height={12} />
              <Text>All videos</Text>
            </StyledMenuItem>
            {playlists.map((item) => (
              <StyledMenuItem
                key={item.id}
                selected={item.name === playlist}
                onClick={() => {
                  onResetFilters();
                  navigateTo("videos", { playlist: item.name });
                }}
                data-testid={`sidebar-item-${item.name}`}
              >
                <Image src="/assets/images/sidebar-item-icon.svg" alt={item.name} width={18} height={12} />
                <Text>{item.name}</Text>
              </StyledMenuItem>
            ))}
          </MenuList>
        )}
        <TagsContainer data-testid="sidebar-tags">
          {tags.map((item) => (
            <Chip
              data-testid={`sidebar-tags-${item.name}`}
              key={item.id}
              onClick={() => {
                onResetFilters();
                navigateTo("videos", { tag: item.name });
              }}
              label={`#${item.name}`}
              variant={tag === item.name ? "filled" : "outlined"}
              size="small"
            />
          ))}
        </TagsContainer>
      </MenuStack>
    </SidebarContainer>
  );
};

export default Sidebar;
