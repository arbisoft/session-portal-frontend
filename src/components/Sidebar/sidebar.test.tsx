import { usePathname, useSearchParams } from "next/navigation";

import useSidebar from "@/hooks/useSidebar";
import { screen, fireEvent, waitFor, customRender } from "@/jest/utils/testUtils";

import Sidebar from "./sidebar";

jest.mock("@/hooks/useSidebar", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  __esModule: true,
  usePathname: jest.fn().mockReturnValue("/some-path"),
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => null), // Mocking no tag selected initially
  })),
}));

const mockNavigateTo = jest.fn();
jest.mock("@/hooks/useNavigation", () => ({
  __esModule: true,
  default: () => ({
    navigateTo: mockNavigateTo,
  }),
}));

describe("Sidebar Component", () => {
  const mockItems = [
    { id: "1", name: "Test-1" },
    { id: "2", name: "Test-2" },
    { id: "3", name: "Test-3" },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
    (useSidebar as jest.Mock).mockReturnValue({
      playlists: mockItems,
      arePlaylistsLoading: false,
      tags: mockItems,
    });
  });

  it("should render Sidebar without crashing", () => {
    customRender(<Sidebar />);
    expect(screen.getByTestId("sidebar-container")).toBeInTheDocument();
  });

  it("should render all sidebar items", () => {
    customRender(<Sidebar />);
    expect(screen.getByTestId("sidebar-item-All")).toBeInTheDocument();
    mockItems.forEach((item) => {
      expect(screen.getByTestId(`sidebar-item-${item.name}`)).toBeInTheDocument();
    });
  });

  it("should highlight selected item", async () => {
    customRender(<Sidebar />);
    const item = screen.getByTestId("sidebar-item-All");

    fireEvent.click(item);

    await waitFor(() => {
      expect(screen.getByTestId("sidebar-item-All")).toHaveStyle("background-color: transparent");
    });
  });

  it("should call navigateTo when a sidebar item is clicked", async () => {
    customRender(<Sidebar />);
    const firstItem = screen.getByTestId("sidebar-item-Test-1");

    fireEvent.click(firstItem);

    await waitFor(() => {
      expect(mockNavigateTo).toHaveBeenCalledWith("videos", { playlist: "Test-1" });
    });
  });

  it("should render sidebar item icons", () => {
    customRender(<Sidebar />);
    expect(screen.getAllByRole("img")).toHaveLength(mockItems.length + 1);
  });

  it("should highlight the selected tag", async () => {
    (usePathname as jest.Mock).mockReturnValue("videos?playlist=Test-1");
    customRender(<Sidebar />);

    await waitFor(() => {
      expect(screen.getByTestId("sidebar-item-Test-2")).toHaveStyle("background-color: transparent");
    });
  });

  it("should allow menu items to be clickable", async () => {
    customRender(<Sidebar />);
    const item = screen.getByTestId("sidebar-item-All");

    fireEvent.click(item);

    await waitFor(() => {
      expect(mockNavigateTo).toHaveBeenCalledWith("videos");
    });
  });

  it("should not change selection when clicking on the same item twice", async () => {
    customRender(<Sidebar />);
    const item = screen.getByTestId("sidebar-item-All");

    fireEvent.click(item);
    fireEvent.click(item);

    await waitFor(() => {
      expect(screen.getByTestId("sidebar-item-All")).toHaveStyle("background-color: transparent");
    });
  });

  it("should support multiple clicks on different items", async () => {
    customRender(<Sidebar />);
    const item1 = screen.getByTestId("sidebar-item-All");
    const item2 = screen.getByTestId("sidebar-item-Test-1");

    fireEvent.click(item1);
    fireEvent.click(item2);
    await waitFor(() => {
      expect(item2).toHaveStyle("background-color: transparent");
    });
  });

  it("should render loading skeleton when data is loading", async () => {
    (useSidebar as jest.Mock).mockReturnValue({ playlists: [], arePlaylistsLoading: true, tags: [] });

    customRender(<Sidebar />);

    await waitFor(() => {
      expect(screen.getByTestId("loading")).toBeInTheDocument();
    });
  });

  it("should match snapshot", () => {
    const { asFragment } = customRender(<Sidebar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render sidebar tags when data is loaded", async () => {
    customRender(<Sidebar />);

    await waitFor(() => {
      expect(screen.getAllByRole("button")).toHaveLength(mockItems.length);
    });
  });

  it("should navigate when clicking on a menu item", async () => {
    customRender(<Sidebar />);
    const item = screen.getByTestId("sidebar-item-Test-2");

    fireEvent.click(item);

    await waitFor(() => {
      expect(mockNavigateTo).toHaveBeenCalledWith("videos", { playlist: "Test-2" });
    });
  });

  it("should show loading state when data is being fetched", async () => {
    (useSidebar as jest.Mock).mockReturnValue({
      arePlaylistsLoading: true,
      playlists: [],
      tags: [],
    });

    customRender(<Sidebar />);
    expect(screen.getByTestId("loading")).toBeInTheDocument();
  });

  it("should allow menu item click when only one item exists", () => {
    (useSidebar as jest.Mock).mockReturnValue({ playlists: [], tags: [], arePlaylistsLoading: false });
    customRender(<Sidebar />);
    const item = screen.getByTestId("sidebar-item-All");
    fireEvent.click(item);
    expect(item).toBeTruthy();
  });

  it("should render all tags", () => {
    customRender(<Sidebar />);
    const tagNames = ["Test-1", "Test-2", "Test-3"];
    tagNames.forEach((name) => {
      expect(screen.getByTestId(`sidebar-tags-${name}`)).toBeInTheDocument();
    });
  });

  it("should highlight the selected tag correctly", () => {
    (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams("tag=2"));
    customRender(<Sidebar />);
    const selectedTag = screen.getByTestId("sidebar-item-Test-2");
    expect(selectedTag).toHaveStyle("background-color: transparent");
  });

  it("should allow clicking on a tag and trigger navigation", () => {
    customRender(<Sidebar />);
    const tag = screen.getByTestId("sidebar-tags-Test-1");
    fireEvent.click(tag);
    expect(mockNavigateTo).toHaveBeenCalledWith("videos", { tag: "Test-1" });
  });
});
