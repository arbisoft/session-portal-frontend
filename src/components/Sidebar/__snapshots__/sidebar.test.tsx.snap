// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-cd9kc0-SidebarContainer-root"
    data-testid="sidebar-container"
  >
    <div
      class="MuiStack-root css-ujfjh4-MuiStack-root-MenuStack-root"
    >
      <ul
        class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-selected MuiMenuItem-root MuiMenuItem-gutters Mui-selected css-dnxijb-MuiButtonBase-root-MuiMenuItem-root-MenuItem-root"
          data-testid="sidebar-item-All"
          role="menuitem"
          tabindex="0"
        >
          <img
            alt="All videos"
            data-nimg="1"
            decoding="async"
            height="12"
            loading="lazy"
            src="/assets/images/sidebar-item-icon.svg"
            style="color: transparent;"
            width="18"
          />
          <p
            class="MuiTypography-root MuiTypography-bodyMedium css-p5v6a9-MuiTypography-root-Text-root"
          >
            All videos
          </p>
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-dnxijb-MuiButtonBase-root-MuiMenuItem-root-MenuItem-root"
          data-testid="sidebar-item-Test-1"
          role="menuitem"
          tabindex="-1"
        >
          <img
            alt="Test-1"
            data-nimg="1"
            decoding="async"
            height="12"
            loading="lazy"
            src="/assets/images/sidebar-item-icon.svg"
            style="color: transparent;"
            width="18"
          />
          <p
            class="MuiTypography-root MuiTypography-bodyMedium css-p5v6a9-MuiTypography-root-Text-root"
          >
            Test-1
          </p>
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-dnxijb-MuiButtonBase-root-MuiMenuItem-root-MenuItem-root"
          data-testid="sidebar-item-Test-2"
          role="menuitem"
          tabindex="-1"
        >
          <img
            alt="Test-2"
            data-nimg="1"
            decoding="async"
            height="12"
            loading="lazy"
            src="/assets/images/sidebar-item-icon.svg"
            style="color: transparent;"
            width="18"
          />
          <p
            class="MuiTypography-root MuiTypography-bodyMedium css-p5v6a9-MuiTypography-root-Text-root"
          >
            Test-2
          </p>
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-dnxijb-MuiButtonBase-root-MuiMenuItem-root-MenuItem-root"
          data-testid="sidebar-item-Test-3"
          role="menuitem"
          tabindex="-1"
        >
          <img
            alt="Test-3"
            data-nimg="1"
            decoding="async"
            height="12"
            loading="lazy"
            src="/assets/images/sidebar-item-icon.svg"
            style="color: transparent;"
            width="18"
          />
          <p
            class="MuiTypography-root MuiTypography-bodyMedium css-p5v6a9-MuiTypography-root-Text-root"
          >
            Test-3
          </p>
        </li>
      </ul>
      <div
        class="MuiBox-root css-1do1o5k-TagsContainer-root"
        data-testid="sidebar-tags"
      >
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault css-mtehtf-MuiButtonBase-root-MuiChip-root"
          data-testid="sidebar-tags-Test-1"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-19m61dl-MuiChip-label"
          >
            #Test-1
          </span>
        </div>
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault css-mtehtf-MuiButtonBase-root-MuiChip-root"
          data-testid="sidebar-tags-Test-2"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-19m61dl-MuiChip-label"
          >
            #Test-2
          </span>
        </div>
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault css-mtehtf-MuiButtonBase-root-MuiChip-root"
          data-testid="sidebar-tags-Test-3"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-19m61dl-MuiChip-label"
          >
            #Test-3
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
